src/Agent.Listener/Configuration/CredentialProvider.cs

```diff
@@ -42,18 +45,21 @@ public CredentialProvider(string scheme)
 
     public sealed class AadDeviceCodeAccessToken : CredentialProvider
     {
-        private string _azureDevOpsClientId = "97877f11-0fc6-4aee-b1ff-febb0519dd00";
+        private VssCredentials _cacheCreds = null;
 
-        public override Boolean RequireInteractive => true;
+        private readonly string _clientId = "97877f11-0fc6-4aee-b1ff-febb0519dd00";
 
+        private readonly string _userImpersonationScope = "499b84ac-1321-427f-aa17-267ca6975798/.default";
         public AadDeviceCodeAccessToken() : base(Constants.Configuration.AAD) { }
 
         public override VssCredentials GetVssCredentials(IHostContext context)
         {
+            if (_cacheCreds != null)
```

Comment:
This will make it impossible to renew the credentials?

instead of this we should probably keep the instance of PublicClientApplication around (that one already handles caching).

Reply:
cache replaced by IPublicClientApplication instance

Reply:
I know it's used only during the setup -- but even then we saw (during the demo) that the setup process asked for the credential twice. Likely once during connection tests and second time during the actual use and agent registration.